src/test/groovy/com/piekoszek/nowaksharedrent/auth/JwtServiceTest.groovy

```diff
@@ -0,0 +1,93 @@
+package com.piekoszek.nowaksharedrent.jwt
+
+import com.piekoszek.nowaksharedrent.jwt.exceptions.InvalidTokenException
+import com.piekoszek.nowaksharedrent.time.TimeService
+import io.jsonwebtoken.SignatureAlgorithm
+import io.jsonwebtoken.security.Keys
+import spock.lang.Specification
+import spock.lang.Subject
+
+import javax.crypto.SecretKey
+
+class JwtServiceTest extends Specification {
+
+    @Subject
+    JwtService jwtService
+
+    TimeService timeService = Mock(TimeService)
+    SecretKey key
+
+    def jwtData = JwtData.builder()
+            .name("exampleName")
+            .email("example@mail.com")
+            .build()
+
+    def setup() {
+        key = Keys.secretKeyFor(SignatureAlgorithm.HS256)
+        jwtService = new JwtServiceConfiguration().jwtService(key, timeService)
+    }
+
+    def "Generate correct token with user name and email"() {
+
+        given: "JWT is generated at second 1 and checked at second 2"
+        timeService.millisSinceEpoch() >>> [1000, 2000]
+        def token = jwtService.generateToken(jwtData)
+
+        expect: "Information saved in and read from the token are the same"
+        jwtService.readToken(token) == JwtData.builder().name("exampleName").email("example@mail.com").build()
+    }
+
+    def "Check valid token"() {
+
+        given: "JWT is generated at second 1 and checked at second 2"
+        timeService.millisSinceEpoch() >>> [1000, 2000]
+        def token = jwtService.generateToken(jwtData)
+
+        when: "Token is checked"
+        jwtService.validateToken(token)
+
+        then: "Nothing happens, which means that token is valid"
+    }
+
+    def "Check expired token"() {
+
+        given: "JWT is generated at second 1 and checked one hour later"
+        timeService.millisSinceEpoch() >>> [1000, 1000 * 60 * 60 + 2000]
+        def token = jwtService.generateToken(jwtData)
+
+        when: "Token is checked"
+        jwtService.validateToken(token)
+
+        then: "Token is expired"
+        thrown(InvalidTokenException)
```

Comment:
Also check exception message, save thrown exception in variable like so:
``def ex = thrown(InvalidTokenException)``
Then check if ex.message is as expected.
Do the same in other test cases.